Added support for withdrawals

BGluth opened the pull request:
Two part PR. The second PR will be upstream in [zero-bin](https://github.com/0xPolygonZero/zero-bin.git).

BGluth committed:
Added support for withdrawals

Nashtare reviewed (approved):
LGTM. As discussed offline, we'll need to possibly update this or plonky2's logic at some point.

Nashtare commented:
Actually @BGluth we may want to think on how to handle from now. I think the reason why we're processing `withdrawals` if there's no txn is to amortize the cost (cc @wborgeaud if I'm wrong). Then we may want to pass an empty vec throughout the entire set of txns, and then always append a dummy proof (in the sense no txn), with the withdrawals, if there are any. 
That way:

- empty blocks aren't impacted, as we keep the `withdrawals` vector empty
- partially empty blocks (1 dummy + 1 txn) and full blocks (2+ txns), may have a light overhead of possibly requiring an additional proof at the end, with no `signed_txn` and the list of `withdrawals`, if it isn't empty. Does that sound alright to you?

wborgeaud commented:
> Actually @BGluth we may want to think on how to handle from now. I think the reason why we're processing `withdrawals` if there's no txn is to amortize the cost (cc @wborgeaud if I'm wrong). Then we may want to pass an empty vec throughout the entire set of txns, and then always append a dummy proof (in the sense no txn), with the withdrawals, if there are any

I guess it depends on the number of withdrawals. Like if there's only one, it might be cheaper to just add the withdrawal to the last txn instead of having an additional proof with an empty txn.
On that topic, the Plonky2 logic is a bit unsound since we should check that the list of withdrawals is empty when a txn is not  the last in the block. (The whole withdrawals logic is unsound tbh since we don't check the withdrawals root, but I guess it's good to keep that in mind)

BGluth commented:
Ah ok, so maybe just hold off on this PR for now then?

Nashtare commented:
We'll need this soon though, if we want to prove real blocks :) But up to you!

Nashtare commented:
@BGluth As we'll need withdrawals to test real-life block, I'd suggest updating the logic to include always a dummy txn for now at the _**end of a block**_, after the last actual txn, containing no txn but the full withdrawals vector.

For partially empty blocks we would then have:
- no difference if there is no withdrawal in the block
- a fully dummy payload and a dummy (no txn) but with the withdrawals vec for the second one in case of a fully empty block with withdrawals
- no more dummy + actual txn but actual txn + dummy payload with withdrawals in case a block contains 1 txn only and withdrawals. 

We can always update / optimize this logic in the future, as William mentioned we could always add them to the last txn if there aren't many, but that will require updating plonky2 anyway, so could be left for future work.

BGluth commented:
Sounds good. I'll update this PR and then give you a ping when it's ready for a review.

dvdplm commented:
Is this PR still relevant?

Nashtare commented:
@dvdplm Yeah, we've been busy with the debugging the whole e2e flow but we still need to support withdrawals

BGluth commented:
Alright I'm going to try to get this done this afternoon. I don't think I made the suggested changed above, so I'll do that now and add @Nashtare as a reviewer.

BGluth committed:
Impled the new suggested way of handling withdrawls

- See
  [this PR post](https://github.com/0xPolygonZero/proof-protocol-decoder/pull/4#issuecomment-1841843807)
  for more info.

BGluth commented:
Ok good to go! Let me know if this is works.

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -281,7 +293,45 @@ impl ProcessedBlockTrace {
+        dummys_added
+    }
+
+    /// The withdrawls are always represented as a single "dummy" txn at the end
```
```suggestion
    /// The withdrawals are always represented as a single "dummy" txn at the end
```

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -281,7 +293,45 @@ impl ProcessedBlockTrace {
+
+        match dummies_already_added {
+            false => {
+                // Guarenteed to have a real txn.
```
```suggestion
                // Guaranteed to have a real txn.
```

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -281,7 +293,45 @@ impl ProcessedBlockTrace {
+                let final_ir = txn_ir.last().unwrap();
+
+                // Dummy state will be the state after the final txn.
+                let withdrawl_dummy =
```
```suggestion
                let withdrawal_dummy =
```

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -281,7 +293,45 @@ impl ProcessedBlockTrace {
+                // of the block.
+                txn_ir.push(withdrawl_dummy);
+            }
+            true => {
+                // If we have dummy txns (note: `txn_ir[1]` is always a dummy txn in this case),
+                // then this dummy will get the withdrawls.
+                txn_ir[1].gen_inputs.withdrawals = withdrawals;
```
I guess this is fine to keep as is, but we may want to emphasize that dummy txns aren't really dummy txns, but more like dummy/empty proofs, as there is no state transition involved. As such, I'd see the withdrawals proof (with no signed_txn) as different from actual padding proofs (the dummy/empty ones currently), but maybe that's just my way of seeing things.
This could probably be rediscussed for a global PR on documentation across the crate though.

Nashtare commented on protocol_decoder/src/processed_block_trace.rs:
```diff
@@ -42,11 +43,16 @@ impl BlockTrace {
         F: CodeHashResolveFunc,
     {
-        let proced_block_trace = self.into_processed_block_trace(p_meta);
+        let proced_block_trace =
```
```suggestion
        let processed_block_trace =
```

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -281,7 +293,45 @@ impl ProcessedBlockTrace {
+        withdrawals: Vec<(Address, U256)>,
+        dummies_already_added: bool,
+    ) {
+        if withdrawals.is_empty() {
```
nit: I'd rather have this condition for calling `add_withdrawls_to_txns`, not inside, but up to you

Nashtare reviewed (changes requested):
Thank you! I wrote some nitpicky comments inline. However, I think we still need some changes on the dummy inputs generation:

- `trie_roots_after` for dummy proofs are constructed as:
```rust
    let trie_roots_after = TrieRoots {
        state_root: tries.state_trie.hash(),
        transactions_root: EMPTY_TRIE_HASH,
        receipts_root: EMPTY_TRIE_HASH,
    };
 ```
 
 This is valid for padding dummy txns, which are always prepended to the list of actual txns in a block, but this won't be for a dummy txn to be created after the entire block execution to accomodate for withdrawals
 
 - The `state_root` component of `tries_root_after` will also need updating for the dummy_txn containing the withdrawals, as these update the state.

BGluth committed:
Apply suggestions from code review

Robin's spelling corrections.

Co-authored-by: Robin Salen <30937548+Nashtare@users.noreply.github.com>

BGluth committed:
Suggested PR changes for #4

BGluth committed:
Withdrawals now adjust account balances

muursh commented on protocol_decoder/src/decoding.rs:
```diff
@@ -43,6 +43,9 @@ pub enum TraceParsingError {
     #[error("Missing keys when creating sub-partial tries (Trie type: {0})")]
     MissingKeysCreatingSubPartialTrie(TrieType),
+
+    #[error("No account present at {0:x} (hashed: {1:x}) to withdrawal {2} Gwei from!")]
```
```suggestion
    #[error("No account present at {0:x} (hashed: {1:x}) to withdraw {2} Gwei from!")]
```

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -268,7 +283,9 @@ impl ProcessedBlockTrace {
         initial_trie_state: &PartialTrieState,
-    ) {
+    ) -> bool {
+        let mut dummies_added = true;
+
```
nit: I think it'd be more "Rust-like" to add a comment explaining what the output boolean is, so that we can directly have 

```rust
match gen_inputs.len() {
    0 => {
        do_something();
        true
    }
    _ => false,
}
```
that returns directly (no intermediary variable)

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -281,8 +298,74 @@ impl ProcessedBlockTrace {
+                txn_ir.push(withdrawal_dummy);
+            }
+            true => {
+                // If we have dummy proofs (note: `txn_ir[1]` is always a dummy txn in this
+                // case), then this dummy will get the withdrawals.
+                txn_ir[1].gen_inputs.withdrawals = withdrawals;
```
I don't think this dummy txn has its `post-state` updated to account for the withdrawals, does it?

Nashtare reviewed (commented):
I think there's still an issue for the case where dummy txns were already added (but state post withdrawal isn't)

dvdplm commented on protocol_decoder/src/decoding.rs:
```diff
@@ -281,8 +298,74 @@ impl ProcessedBlockTrace {
+        dummies_already_added: bool,
+    ) -> TraceParsingResult<()> {
+        // Withdrawals update balances in the account trie, so we need to do that here.
+        Self::update_trie_state_from_withdrawals(withdrawals.iter(), &mut final_trie_state.state)?;
```
How about making `update_trie_state_from_withdrawals` take `impl IntoIterator<… …>` instead? Sp you can avoid this `.iter()`?

BGluth committed:
Suggested PR changes for #4 (2)

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -264,26 +279,107 @@ impl ProcessedBlockTrace {
+                // If we have dummy proofs (note: `txn_ir[1]` is always a dummy txn in this
+                // case), then this dummy will get the withdrawals.
+                txn_ir[1].gen_inputs.withdrawals = withdrawals;
+                txn_ir[1].gen_inputs.tries.state_trie = final_trie_state.state.clone();
```
sorry I wasn't clear! Only `trie_roots_after` should be updated with the execution of the withdrawals. The `tries` (i.e. before) should match the final state after the last transaction. Here we have the same `state` trie/root, as the `final_trie_state` we're copying as been mutated to include withdrawals already.

BGluth committed:
Suggested PR changes for #4 (3)

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -264,26 +279,107 @@ impl ProcessedBlockTrace {
+
+                // Dummy state will be the state after the final txn.
+                let withdrawal_dummy =
+                    create_dummy_gen_input(other_data, final_trie_state, final_ir.txn_idx + 1);
```
same here, we are having identical state for before and after for this dummy. While it should be 

initial: final_trie_state
// update final_trie_state with withdrawals
final: final_trie_state

BGluth committed:
Suggested PR changes for #4 (4)

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -264,26 +279,119 @@ impl ProcessedBlockTrace {
+                withdrawal_dummy.gen_inputs.trie_roots_after.state_root =
+                    final_trie_state.state.hash();
+
+                // If we have no actual dummy proofs, then we create one and append it to the
+                // end of the block.
```
nit: I think it may be better at the start of this match bracket?

Nashtare commented on protocol_decoder/src/decoding.rs:
```diff
@@ -264,26 +279,119 @@ impl ProcessedBlockTrace {
+                    &mut final_trie_state.state,
+                )?;
+
+                // Only the state root hash needs to be updated from the withdrawals.
+                withdrawal_dummy.gen_inputs.trie_roots_after.state_root =
+                    final_trie_state.state.hash();
```
We should also add the withdrawals:

```rust
withdrawal_dummy.gen_inputs.withdrawals = withdrawals;
```

BGluth committed:
Suggested PR changes for #4 (5)

Nashtare reviewed (approved):
LGTM :)
